packages/protocol/contracts/governance/Election.sol

```diff
@@ -0,0 +1,779 @@
+pragma solidity ^0.5.3;
+
+import "openzeppelin-solidity/contracts/math/Math.sol";
+import "openzeppelin-solidity/contracts/math/SafeMath.sol";
+import "openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "openzeppelin-solidity/contracts/utils/ReentrancyGuard.sol";
+
+import "./interfaces/IValidators.sol";
+import "../common/Initializable.sol";
+import "../common/FixidityLib.sol";
+import "../common/linkedlists/AddressSortedLinkedList.sol";
+import "../common/UsingRegistry.sol";
+
+
+contract Election is Ownable, ReentrancyGuard, Initializable, UsingRegistry {
+
+  using AddressSortedLinkedList for SortedLinkedList.List;
+  using FixidityLib for FixidityLib.Fraction;
+  using SafeMath for uint256;
+
+  // Pending votes are those for which no following elections have been held.
+  // These votes have yet to contribute to the election of validators and thus do not accrue
+  // rewards.
+  struct PendingVotes {
+    // Maps groups to total pending voting balance.
+    mapping(address => uint256) total;
+    // Maps groups to accounts to pending voting balance.
+    mapping(address => mapping(address => uint256)) balances;
+    // Maps groups to accounts to timestamp of the account's most recent vote for the group.
+    mapping(address => mapping(address => uint256)) timestamps;
+  }
+
+  // Active votes are those for which at least one following election has been held.
+  // These votes have contributed to the election of validators and thus accrue rewards.
+  struct ActiveVotes {
+    // Maps groups to total active voting balance.
+    mapping(address => uint256) total;
+    // Maps groups to accounts to the numerator of the account's fraction of the group's
+    // total active votes.
+    mapping(address => mapping(address => uint256)) numerators;
+    // Maps groups to the denominator of all accounts' fraction of the group's total active votes.
+    mapping(address => uint256) denominators;
+  }
+
+  struct TotalVotes {
+    // The total number of votes cast.
+    uint256 total;
+    // A list of eligible ValidatorGroups sorted by total votes.
+    SortedLinkedList.List eligible;
+  }
+
+  struct Votes {
```

Comment:
After reading the whole contract, my impression is that we have some field that have a derived value, and we should:
1) review if we need that
2) have a clear definition of what's the derivation rule.

For example `TotalVotes.total` is related to `eligible` but that relation seems to be sometimes broken.

Also, I think we can structure state differently, following to what @m-chrzan suggested. There's a clear: `Map<group, Map<address, XXX>>` where `XXX` has pending balances, pending timestam, active numerator.

Reply:
yes! i like this much better...
maybe we can remove TotalVotes, since it only has one field now.